### PR TITLE
REPL/LineEdit to support "undo": part of #8447

### PR DESCRIPTION
**Pull request by srp**
The REPL now supports undo via Ctrl-^ or Ctrl-_.

This should very closely minic the behavior of other readline/emacs
shells, except it doesn't let the user goto a historical entry (A),
edit it, goto a different historical entry (B), return to the first
(A) and undo the edits.

---

This change is [Reviewable](https://reviewable.io/reviews/julialang/julia/9596)

### CODE REVIEW

**Comment by rfourquet** [comment]
Thanks! I tested it and works fine. I don't know the LineEdit.jl code very well yet, but the patch looks good to me.
WRT the limitation you mention in the commit: would it be difficult to eventually attach an `undo_buffers` vector to each history entry? (but it's maybe not worth the effort, I don't think I use this functionality in readline). 
There is also another difference from readline, which packs together individual characters inserts (i.e. typing "123" and then undo removes those 3 characters), but I don't see this as a problem.

File: base/LineEdit.jl
Comment (rfourquet) [inline base/LineEdit.jl]: Why not `IOBuffer[]` (w/o parens) ?

**Comment by rfourquet** [comment]
Oh, and special +1 for extensive tests.
If you didn't start yet implementing beginning/end-of-history from #8447, I have it here and will do a PR soon.

**Comment by srp** [comment]
Looking forward to the beginning/end-of-history! My plan was to next tackle kill/yank which doesn't seem to work quite right: m-d and friends don't get added to yank ring, yank ring isn't yet a ring, etc.

**Comment by rfourquet** [comment]
Bump. I'm really missing this functionality!

File: base/LineEdit.jl
Comment (blakejohnson) [inline base/LineEdit.jl]: why have the calls to `seek` been dropped here and elsewhere in this PR?

Comment (rfourquet) [inline base/LineEdit.jl]: Probably because `edit_replace` already is "undo aware", whereas `seek` is not. When I compile without this change, after an undo the cursor is misplaced (e.g. at the beginning of a (prefix) word after undoing a completion).

**Comment by rfourquet** [comment]
I rebased locally this branch on master, and for some reason the key-binding "^/" is not valid anymore (compilation fails), I don't know why yet.

**Comment by rfourquet** [comment]
@srp would you consider rebasing this work to get it merged?

**Comment by StefanKarpinski** [comment]
Yes, please – this would be a lovely feature to have. Sorry the PR has languished!

File: base/repl/LineEdit.jl
Comment (tkelman) [inline base/repl/LineEdit.jl]: this line should probably be wrapped

Comment (srp) [inline base/repl/LineEdit.jl]: done

**Comment by srp** [comment]
@rfourquet,@StefanKarpinski: rebased, thanks!

File: base/repl/LineEdit.jl
Comment (rfourquet) [inline base/repl/LineEdit.jl]: maybe you could `push_undo` only if the buffer is not empty already? Another possibility could also be to not push a new state in `push_undo` when it's equal to the last pushed state?

File: base/repl/LineEdit.jl
Comment (rfourquet) [inline base/repl/LineEdit.jl]: maybe you could `push_undo` only if `!isempty(s.kill_buffer)`?

File: base/repl/LineEdit.jl
Comment (rfourquet) [inline base/repl/LineEdit.jl]: `push_undo` only if some characters are deleted?

File: base/repl/LineEdit.jl
Comment (rfourquet) [inline base/repl/LineEdit.jl]: `isempty(s.undo_buffers) && return false`

File: test/lineedit.jl
Comment (rfourquet) [inline test/lineedit.jl]: Use `@testset` ?

File: doc/src/manual/interacting-with-julia.md
Comment (rfourquet) [inline doc/src/manual/interacting-with-julia.md]: I didn't see you add a key for `^/`, how does it work? (when I tried to rebase last year, I had compilation problems with this key...)
Also, your commit message refers to "Ctrl-^", you should change it.

File: base/repl/LineEdit.jl
Comment (rfourquet) [inline base/repl/LineEdit.jl]: I would prefer the names of the new functions with an ending `!`, but it seems to be the style of the file, so better to keep as is for now.

**Review by rfourquet — APPROVED** [review]
Thanks for your very nice work! None of my comments are blocking, and can be addressed later if needed.
I will merge in a few days if no-one has objections.

**Comment by KristofferC** [comment]
How do I type `Ctrl+/` on a European (Swedish) keyboard? Would `Ctrl + Shift + 7` work where `Shift + 7` gives `/`?

**Comment by fredrikekre** [comment]
`Ctrl + Shift + 7` works for me @KristofferC :sweden: 
I have to say that it is not a very convenient keyboard short cut for such keyboard layouts though.

File: base/repl/LineEdit.jl
Comment (rfourquet) [inline base/repl/LineEdit.jl]: Also, this one is a good candidate for short-form function syntax.

File: doc/src/manual/interacting-with-julia.md
Comment (srp) [inline doc/src/manual/interacting-with-julia.md]: ^/ isn't a valid control character (eg after subtracting 64 you get a negative number), however it appears that vt102 terminal emulators map it to emit ^_ for convience, see: https://apple.stackexchange.com/a/227286

From the terminals I tested on this is accurate that pressing ctrl-/ causes "^_" to appear in the console.

**Comment by srp** [comment]
Does Ctrl+_ work on a Swedish keyboard? If so, you might have to stick with that.

**Comment by fredrikekre** [comment]
`_` is `Shift + -`

**Comment by srp** [comment]
Sounds the same as an english keyboard, i have to hit `Ctrl + Shift + -` to emit ^_

**Comment by srp** [comment]
@rfourquet I plan on pushing an update addressing some of your comments, but in the process noticed that undo doesn't work right (any more?) with bracketed insert. I'm trying to fix that, but so far it's not entirely straight forward.

**Comment by rfourquet** [comment]
@srp, I rebased your branch and tried to fix undo with the bracketed paste mode, would you mind checking out this [branch](https://github.com/JuliaLang/julia/tree/srp/repl-undo) (just do `git checkout srp/repl-undo && make` after fetching from the julia repo) and verify if this fixes your concerns? You can use this to update your branch (which now has a small conflict). I wish to merge your work soon!

**Comment by rfourquet** [comment]
Bump - I have few features in the pipeline which will increase the work to resolve conflicts here if we don't merge this first...

File: base/repl/LineEdit.jl
Comment (ararslan) [inline base/repl/LineEdit.jl]: These should probably all be refactored to have `!`. It wasn't clear to me reading through this that it actually modified the `PromptState` until I saw the definition. The larger refactoring doesn't have to happen in this PR, but if you're adding a function now, might as well use the standard notation, IMO.

**Comment by rfourquet** [comment]
I rebased again and updated the branch "srp/repl-undo" that I linked to above. I plan to merge by tomorrow. But I cannot update the PR directly, and I'm afraid that if I merge it on the command line, github won't recognize this PR has been merged (it happened to me already), someone would have a hint as to how to do it properly?

**Comment by rfourquet** [comment]
Thanks again @srp !

**Pull request merged**
